Make handshake backward-compatible

blampe opened the pull request.

blampe commented on sdk/go/common/resource/plugin/provider_server.go:
```diff
@@ -135,7 +136,12 @@ func (p *providerServer) Handshake(
-	_, err := p.provider.Handshake(ctx, ProviderHandshakeRequest{
+	h, ok := p.provider.(handshaker)
+	if !ok {
+		return nil, errors.New("Handshake not implemented")
```
Alternatively

```go
return &pulumirpc.ProviderHandshakeResponse{}, nil
```

Frassle commented:
We'd be better off moving this interface to be like the grpc interface where you need to embed a "UnimplementedProvider" or "UnsafeProvider" struct. Better than adding a new interface everytime we add a method on the grpc layer (we'd be up to about five interfaces now if we'd been doing that).

But also we _really_ ought to just write up a new clean version of Provider that is clearly part of the public api (i.e. not sdk/go/common) and make that like the grpc interfaces (i.e. context.Context everywhere, struct args, unimplemented struct).

blampe commented:
> Better than adding a new interface everytime we add a method on the grpc layer (we'd be up to about five interfaces now if we'd been doing that).

Why is that a problem? Smaller, composable interfaces are ideal!

Frassle reviewed (changes requested):
Ok so my memory forgot Ian had already done most of what I'd said above (we still ought to move this out of sdk/go/comon though). This interface is _already_ a mustEmbed style interface (https://github.com/pulumi/pulumi/blob/14fd3155ce43bcff9e8b821b4b00e7689c49d099/sdk/go/common/resource/plugin/provider.go#L387). 

Its just the bridge and the service have opt-ed out of forward compatibility: 
https://github.com/pulumi/pulumi-terraform-bridge/blob/b337517ee358b683f7a9fbc89b189d8ae5734fbb/pkg/tfbridge/provider.go#L93

https://github.com/pulumi/pulumi-service/blob/fdb1a163a8cb74515d9550ba768dc0ed89021e98/cmd/pulumi-insights-scanner/main_test.go#L32

This is an issue in those repos. Nothing needs changing here, if you want service and bridge to get clean upgrades then opt-in that. For the engine it's more useful for us to get notified of every place that needs updating when we add a new method.

> Why is that a problem? Smaller, composable interfaces are ideal!

They also increase the number of named things, the number of runtime type tests, are less discoverable, and as above don't give us compiler errors to fix up definition sites that _should_ be fixed up when we add them.

Hopefully that's all enough explanation to understand why I don't support this change.

blampe commented:
> This interface is _already_ a mustEmbed style interface

Why though? Interfaces exist to let each side of an API contract evolve without tight coupling. Why do we want to _force_ coupling like this?

> This is an issue in those repos. Nothing needs changing here, if you want service and bridge to get clean upgrades then opt-in that.

It's not so simple. Take a look at the service.

```go
type mockProvider struct {
    plugin.NotForwardCompatibleProvider
    *mocks.MockProvider
}
```

Insights is generating `MockProvider` from the `plugin.Provider` interface. What they're _trying_ to do is [mock](https://github.com/uber-go/mock) some methods -- `List` and `GetSupportedTypes` -- so they can unit test their logic. Very reasonable! We want to encourage more of this kind of testing.

```go
	lister.EXPECT().GetSupportedTypes().Return([]insights.ListResourceType{
		{TypeToken: "aws:ec2:Vpc", DependentTypes: nil},
		{TypeToken: "aws:s3:Bucket", DependentTypes: []insights.ListResourceType{{TypeToken: "aws:s3:BucketObject", DependentTypes: nil}}},
	})
```

They try opting-in to forward compatibility:
```go
type mockProvider struct {
    plugin.UnimplementedProvider
    *mocks.MockProvider
}
```

But this doesn't work, because `UnimplementedProvider` and `MockProvider` both implement the same methods so there's a collision. They also can't use `mocks.MockProvider` directly, because it was generated from the older interface. So what are their options?

1. Re-generate the mock from the new interface. This is straightforward _but it means no forward-compatibility_. No good.
2. Start manually maintaining a mock that embeds `UnimplementedProvider`. Yikes! We don't want to force them to throw away their tools.

> For the engine it's more useful for us to get notified of every place that needs updating when we add a new method.

I _totally_ get that, the compiler is king and I'm not saying we need to get rid of the "full" `Provider` interface. I'm just saying it shouldn't be part of the public API. The engine should be able to accept a provider that doesn't implement every method -- and yes that's the case today but we've accomplished it in a very roundabout way with `UnimplementedProvider`.

(For what it's worth, I only see `MockProvider`, `provider`, and `UnimplementedProvider` as needing updating for new methods -- and the latter becomes unnecessary.)

We already _sort of_ do what I'm describing with `Attach`:

https://github.com/pulumi/pulumi/blob/4ab224fd702aa523ea4fd18a730313de6cb265f5/sdk/go/common/resource/plugin/provider_server.go#L210-L218

More concretely, what I'm suggesting is:
* The kitchen-sink `Provider` can continue to exist but not as input to `NewProviderServer`. I'd recommend `NewProviderServer(p any)` and test for methods individually, but you could also freeze `Provider` as (e.g.) `NewProviderServer(p ProviderAsOf202412)` and expand that similarly to what we did for `Attach`. 
* For the `NotForwardCompatibleProvider` use cases that need to guarantee the full interface is satisfied, do a more idiomatic compliance check instead: `var _ Provider = (*MockProvider)(nil)`.

Taken together those give you forward-compatibility _or_ strictness without forcing any decisions on how consumers use your types.

Frassle commented:
> Why though? Interfaces exist to let each side of an API contract evolve without tight coupling. Why do we want to force coupling like this?

mustEmbed is supposed to be best of both worlds. The clients that want to see the break changes opt-in via `NotForwardCompatibleProvider`, those that don't want the tight coupling can use `UnimplementedProvider`.

> We already sort of do what I'm describing with Attach:

Attach was from many years ago before we worked out that copying grpc and using "mustEmbed" seemed a sensible way to evolve this interface. We also haven't copied that approach with any of the other new methods since then (GetMapping, GetMappings, Parameterize, Handshake).

> What they're trying to do is [mock](https://github.com/uber-go/mock) some methods --

Sigh this feels like two good practices hitting heads. 
It kind of feels like a gomock issue. Your asking gomock to mock the provider interface and it's failing at it because the type it generates `MockProvider` does not implement the `Provider` interface because of the need to have a nested "mustEmbed" struct.

I'd say it's worth raising an issue with gomock about this. Given this is _also_ how grpc works I'd hope there could be some support for this code style from gomock.

Failing that I'd say it's a fight of what's more costly. Service needing to regen mocks at some cadence (this interface doesn't get new stuff that often), or just editing the gomock file to add `plugin.UnimplementedProvider` when they do regen it (*). Or platform needing to name and maintain more interfaces. I know that latter looks like the easy choice _now_ but considering if we'd made a decision like this three years ago we'd now have `Provider`, `ProviderWithAttach`, `ProviderWithGetMapping`, `ProviderWithGetMappings`, `ProviderWithParameterize`, and all the requisite type checks throughout the engine to deal with those. Which I think gets back to...

> I'm just saying it shouldn't be part of the public API

Aye! I've been saying we should do a new "best practice" stable public api interface for providers outside of sdk/go/common for ages now. If best practice is an interface per new method we can consider that if it's just the public interface code for defining a provider server, and not affecting typing through the rest of the engine system. Although I'd still question if that's _really_ the best way, or if the better fix here is gomock supporting this style of construct.


(*)
It requires editing a generated file, but it does cleanup the test file well:
```
fraser@IBASA-PC:~/projects/pulumi-service$ git diff
diff --git a/cmd/pulumi-insights-scanner/main_test.go b/cmd/pulumi-insights-scanner/main_test.go
index f2981084b..16b610a52 100644
--- a/cmd/pulumi-insights-scanner/main_test.go
+++ b/cmd/pulumi-insights-scanner/main_test.go
@@ -28,13 +28,6 @@ import (
        "go.uber.org/mock/gomock"
 )
 
-type mockProvider struct {
-       plugin.NotForwardCompatibleProvider
-       *mocks.MockProvider
-}
-
-var _ = plugin.Provider(mockProvider{})
-
 func TestScan(t *testing.T) {
        c := gomock.NewController(t)
        escClient := mocks.NewMockESCClient(c)
@@ -88,7 +81,7 @@ func TestScan(t *testing.T) {
                resource.PropertyMap{
                        "region": resource.NewStringProperty("us-west-2"),
                },
-       ).Return(lister, mockProvider{MockProvider: provider}, pkgSchema, nil)
+       ).Return(lister, provider, pkgSchema, nil)
 
        lister.EXPECT().GetSupportedTypes().Return([]insights.ListResourceType{
                {TypeToken: "aws:ec2:Vpc", DependentTypes: nil},
@@ -248,7 +241,7 @@ func TestConfigOverride(t *testing.T) {
                "region":      resource.NewStringProperty(expectedRegion),
                "newProperty": resource.NewNumberProperty(123),
        },
-       ).Return(lister, mockProvider{MockProvider: provider}, pkgSchema, nil)
+       ).Return(lister, provider, pkgSchema, nil)
 
        logger := slog.New(slog.NewJSONHandler(io.Discard, nil))
        scanner := newScanner(logger, scan, escClient, host)
diff --git a/cmd/pulumi-insights-scanner/mocks/provider.go b/cmd/pulumi-insights-scanner/mocks/provider.go
index f6776455a..263254dd5 100644
--- a/cmd/pulumi-insights-scanner/mocks/provider.go
+++ b/cmd/pulumi-insights-scanner/mocks/provider.go
@@ -21,6 +21,7 @@ import (
 
 // MockProvider is a mock of Provider interface.
 type MockProvider struct {
+       plugin.UnimplementedProvider
        ctrl     *gomock.Controller
        recorder *MockProviderMockRecorder
 }
```
And if you forget to remake the edit when you regenerate you get compile errors.

blampe commented:
> Sigh this feels like two good practices hitting heads.

My point is that `mustEmbed` is _not_ a best practice -- it's an anti-pattern. Generated gRPC/proto code is _bad_ and not something to emulate.

> The clients that want to see the break changes opt-in via NotForwardCompatibleProvider, those that don't want the tight coupling can use UnimplementedProvider.

Again, we already get this behavior out-of-the-box without needing embeds. If the engine upcasts new methods (as it did with `Attach`) our consumers get backwards compatibility by default. The clients that want to see breaking changes (effectively just the bridge) can already opt-in to that with `var _ Provider = (*provider)(nil)`. So what do these embeds buy us?

> if we'd made a decision like this three years ago we'd now have Provider, ProviderWithAttach, ProviderWithGetMapping, ProviderWithGetMappings, ProviderWithParameterize, and all the requisite type checks throughout the engine to deal with those.

That's just the cost of evolving APIs and maintaining backwards compatibility, though. It's not much trouble at all once the convention is established.

(Worth mentioning these would be private interfaces if your concern is about public surface area.)

Frassle commented:
> My point is that mustEmbed is not a best practice -- it's an anti-pattern. Generated gRPC/proto code is bad and not something to emulate.

Apart from gomock this approach has served us well. As far as we we're all aware it seemed a good way of dealing with evolving interfaces. I'll give that the gomock issues are compelling, although I think `mockgen --embed plugin.UnimplementedProvider` would probably be a good fix as well.

> So what do these embeds buy us?

I've said this a lot above but to repeat again. They buy us compile time checks, no requirement for runtime upcasts, less named objects, and more discoverable methods. Apart from the fact gomock doesn't like them they seemed better than adding a new named interface for each method.

> That's just the cost of evolving APIs and maintaining backwards compatibility, though. It's not much trouble at all once the convention is established.

Well it's _not_ been the cost so far because this embed trick has been working out. If we want to take this path the convention does need to be established.

> (Worth mentioning these would be private interfaces if your concern is about public surface area.)

I don't think they can be private interfaces. We want these to be in the public documentation when you look at the package docs, it's already going to less discoverable that all these interfaces are actually methods for one provider object than just having them on the one provider interface, we can't make that even worse by making them internal and hidden. It also means we can have a way to opt-in to a "give me everything" interface for the engine and any downstreams who do care about getting compile time errors on change.

My current thinking to approach this so far, although it still feels a bit unsure, is four parts.
1. Lock the provider interface as it is now but as a _new_ interface called something like "BaseProvider".
2. Remove mustEmbed, UnimplementedProvider, NotForwardCompatibleProvider as they won't be used anymore.
3. Any new methods get added as a new interface, named something like "ProviderWithHandshake". This feels like the most important part to set a good convention around.
4. Rewrite the current "Provider" interface as an interface that just inherits "BaseProvider" and all the new interfaces added by step 3.

The engine can then continue using Provider and get its compile time safety. Downstream users who want forward-compatible interfaces can just use "BaseProvider" (or define a new interface with the additions they care about). Users can still get public documentation about provider methods, even if not all nicely on one clear interface.

I think the only big decision is clearly deciding and documenting what the convention for part three is. If we want to retro-actively move handshake into that style (even though by this point everyone's updated) we could.

This will need clearly documenting in the source code though. It'll be _far_ too easy to forgot.

Frassle commented:
Why is it always _after_ you write something down your brain kicks into gear to think through more implications 🙄 

Thinking this over, if we still want to support something like `UnimplementedProvider` (which I think we do, given it's uses in testing) I think we'd have to split "Provider" up into an interface per-method now or keep `UnimplementedProvider` as is to allow users to create test types that satisfy "Provider" interface without needing to write each method. That does lead to this odd case of kinda supporting both worlds, split interfaces for new things, "UnimplementedProvider" for default implementations of old and new things.

blampe commented:
> I don't think they can be private interfaces. We want these to be in the public documentation when you look at the package docs, it's already going to less discoverable that all these interfaces are actually methods for one provider object than just having them on the one provider interface, we can't make that even worse by making them internal and hidden. It also means we can have a way to opt-in to a "give me everything" interface for the engine and any downstreams who do care about getting compile time errors on change.

They're still discoverable on the "give me everything" interface -- I updated the PR with `UnsafeProvider` as an example. The docs for that will cover all of the supported methods. 

I agree with your suggestion to rename to BaseProvider but added this for the sake of keeping the example more self-contained. I would suggest keeping "Unsafe" or something like that as part of the name to make it more obvious it's not for implementations.

> Remove mustEmbed, UnimplementedProvider, NotForwardCompatibleProvider as they won't be used anymore.

They should be marked deprecated instead of removed, but yeah.

> Thinking this over, if we still want to support something like UnimplementedProvider (which I think we do, given it's uses in testing) I think we'd have to split "Provider" up into an interface per-method now or keep UnimplementedProvider as is to allow users to create test types that satisfy "Provider" interface without needing to write each method. That does lead to this odd case of kinda supporting both worlds, split interfaces for new things, "UnimplementedProvider" for default implementations of old and new things.

The idiomatic way to do this would break out all of these methods, so you your tests only need to implement the method(s) they care about. It's straightforward to do that with the existing methods, and that's backward-compatible.
